Implement new side navigation menu

miguelaenlle opened the pull request:
Implementation of #11229. Note that all `PageLayout` changes in this PR were merged from #11324.

The new navigation menus can be enabled from **Global Admin > Features**:

![Screenshot 2025-02-13 at 11 39 47 PM](https://github.com/user-attachments/assets/99414a21-bdaa-4a28-a78f-501ce19f80ce)

Here's what the new navigation menus look like from the following pages:

**Global admin:**

![image](https://github.com/user-attachments/assets/cc6a6f94-a86d-44aa-b1b9-076c91d1365d)

**Institution admin:**

![image](https://github.com/user-attachments/assets/b8aca161-af4a-498d-b7d5-6c861ae6595f)

**Course:**

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/918bc548-dd87-4f1a-a8d3-50d3613b628a" />

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/1f12e88e-40a8-4855-84c0-ab0953027f9a" />

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/58e4c054-f554-4523-9a61-2e95efce0333" />

**Course instance:**

<img width="1710" alt="Screenshot 2025-02-13 at 11 42 59 PM" src="https://github.com/user-attachments/assets/3c7be3b2-f388-453e-9e4b-1276cc1a57b2" />

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/75dd55ba-df06-4bd2-9462-4842b871b4e5" />

**Assessment:**

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/97c57ef8-b559-4c31-8cfc-8cd0acc8581e" />

<img width="1710" alt="image" src="https://github.com/user-attachments/assets/94097617-0e31-4c79-b721-af9b83b83710" />

The top navbar buttons redirect to the following pages:

<img width="697" alt="image" src="https://github.com/user-attachments/assets/9be0f799-5c98-47fd-93de-9e7d8edcd6fb" />

- Home -> `/`
- Global Admin -> `/pl/administrator/admins`
- Institutions -> `/pl/administrator/institutions`
- [Institution Name] -> `/pl/administrator/institution/[Institution ID]`
- Courses -> `/pl/administrator/institution/[Course ID]/courses`
- [Course Name] -> `/pl/course/[Course ID]/course_admin/instances`

Information displayed on the top navbar changes based on the user's access level.

**Global admin:**

<img width="787" alt="image" src="https://github.com/user-attachments/assets/3948550f-c441-4a72-a10f-f32f225de41e" />


**Institution admin:**

<img width="553" alt="image" src="https://github.com/user-attachments/assets/da539b6f-75ae-4a6a-91ff-da6da027b69a" />


**Course admin/course instance admin:**

<img width="463" alt="image" src="https://github.com/user-attachments/assets/ea6de60a-f2e6-4383-a425-c03f338ee025" />


Features to add in future PRs:
- Mobile optimization
- Ability to toggle the side nav bar

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -612,248 +757,267 @@ function NavbarInstructor({
+      { text: 'Courses', href: `/pl/administrator/institution/${institution.id}/courses` },
+      { text: course.short_name, href: `/pl/course/${course.id}/course_admin/instances` },
+    ],
+    options: {
```
Remove these options in the consolidated navbarButtons

miguelaenlle commented on apps/prairielearn/src/components/NavbarContext.html.ts:
```diff
@@ -14,419 +15,478 @@ interface TabInfo {
   resLocals,
   navPage,
   navSubPage,
+  hasEnhancedNavigation = false,
```
This takes in resLocals, so is hasenhacnedNavigation needed?

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -43,40 +47,137 @@ export function PageLayout({
-          class="
+  if (resLocals.has_enhanced_navigation) {
+    // The left navbar is only shown if the user is in a
+    // course or course instance page.
```
page within a course or course instance

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -43,40 +47,137 @@ export function PageLayout({
+    if (navContext.page === 'course_admin') {
+      if (
+        navContext.subPage &&
+        ['settings', 'sets', 'modules', 'tags', 'topics', 'staff'].includes(navContext.subPage)
```
Explain why these in particular. 
Also, can you retrieve these from their actual list?

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -43,40 +47,137 @@ export function PageLayout({
+                <main
+                  id="content"
+                  class="
+                ${options.fullWidth ? 'container-fluid' : 'container'} 
```
Can you clean up this formatting?

miguelaenlle commented on apps/prairielearn/src/ee/pages/administratorInstitutionCourses/administratorInstitutionCourses.html.ts:
```diff
@@ -1,7 +1,6 @@
 import { html } from '@prairielearn/html';
```
Apply these changes to all the institution admin pages, and put them in a separate PR

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,343 @@
+  /**
+   * TODO: complete this comment
+   **/
+  checkActiveSubPageForPages?: NavPage[];
```
move this before activeSubPage

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,343 @@
+    {
+      activePage: ['instance_admin', 'assessments', 'assessment', 'assessment_instance'],
+      activeSubPage: ['assessments'],
+      checkActiveSubPageForPages: ['instance_admin'],
```
Nit: move this up one line

miguelaenlle commented on apps/prairielearn/src/components/NavbarContext.html.ts:
```diff
@@ -14,419 +15,478 @@ interface TabInfo {
+
+  const navPagesTabs = getNavPageTabs(hasEnhancedNavigation);
   const navPageTabs = navPagesTabs[navPage];
+  const showLeftNavbar = resLocals.course !== undefined;
```
Change all showLeftNavbar instances to showSideNav

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,343 @@
+
+  let isActive = navPage !== null && activePage.includes(navPage);
+  if (isActive && (!checkActiveSubPageForPages || checkActiveSubPageForPages.includes(navPage))) {
+    isActive = navSubPage != null && activeSubPage.includes(navSubPage);
```
Why is navSubPage not required?

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,343 @@
+  const urlSuffix =
+    typeof tabInfo.urlSuffix === 'function' ? tabInfo.urlSuffix(resLocals) : tabInfo.urlSuffix;
+
+  let isActive = navPage !== null && activePage.includes(navPage);
```
Why is navPage not required?

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,343 @@
+  `;
+}
+
+// return html`
```
Remove these comments

miguelaenlle commented on apps/prairielearn/src/pages/instructorAssessments/instructorAssessments.html.ts:
```diff
@@ -56,155 +55,161 @@ export function InstructorAssessments({
-              <h1>Assessments</h1>
-              ${authz_data.has_course_permission_edit && !course.example_course && rows.length > 0
-                ? html`
+  return PageLayout({
```
Move these changes to the separate PR?
Also, implement PageLayout for AI question generation

miguelaenlle commented on apps/prairielearn/src/server.ts:
```diff
@@ -1283,43 +1284,28 @@ export async function initExpress(): Promise<Express> {
       next();
     }),
   );
-  app.use('/pl/course_instance/:course_instance_id(\\d+)/instructor/instance_admin/settings', [
```
Check these after PageLayout PRs

miguelaenlle replied:
Error pages are missing top padding

miguelaenlle replied:
http://localhost:3000/pl/course/13/effectiveUser missing PageLayout

miguelaenlle reviewed (commented):
Initial self-PR.

miguelaenlle commented on apps/prairielearn/src/components/AssessmentNavigation.html.ts:
```diff
@@ -0,0 +1,40 @@
+import { html } from '@prairielearn/html';
+
+import type { Assessment, CourseInstance } from '../lib/db-types.js';
+
```
Add a comment describing what this is

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -119,20 +126,32 @@ function NavbarByType({
-  if (navbarType == null || navbarType === 'plain') {
-    return NavbarPlain({ resLocals, navPage });
-  } else if (navbarType === 'student') {
+  if (navbarType === 'student') {
```
Explain the rationale behindd doing this

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -119,20 +126,32 @@ function NavbarByType({
+      } else if (navbarType === 'institution') {
+        return NavbarInstitution({ resLocals });
+      } else if (navbarType === 'public') {
+        return NavbarPublic({ resLocals });
```
Why is this here?

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -567,13 +586,117 @@ function AuthnOverrides({
+  navPage: NavPage;
+  navbarType: NavbarType;
+}) {
+  if (
```
Explain why we do this

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -38,45 +44,150 @@ export function PageLayout({
   /** The main content of the page within the main container. */
   content: HtmlValue;
-  /** The content of the page in the body after the main container. */
+  /** The content of the page in the body after the min container. */
```
Inaccurate comment

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -38,45 +44,150 @@ export function PageLayout({
+    const showSideNav =
+      navContext.type !== 'student' &&
+      navContext.type !== 'public' &&
+      resLocals.course !== undefined;
```
&& resLocals.course

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -38,45 +44,150 @@ export function PageLayout({
+
+    // ContextNavigation is shown if no left navbar is shown or
+    // additional navigation capabilities are needed alongside the
+    // left navbar
```
Provide an example

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -38,45 +44,150 @@ export function PageLayout({
+    // left navbar
+    if (navContext.page === 'course_admin') {
+      const navPageTabs = getNavPageTabs(true);
+      // On the course admin page, these are the sub pages that
```
Improve comments

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,338 @@
+import { ProgressCircle } from './ProgressCircle.html.js';
+
+interface SideNavTabInfo {
+  /** We must be on activePage for the tab to be active.  */
```
Improve comments

miguelaenlle reviewed (commented):
Second self-PR.

miguelaenlle commented on apps/prairielearn/src/ee/pages/administratorInstitutionAdmins/administratorInstitutionAdmins.html.ts:
```diff
@@ -11,26 +10,22 @@ export function AdministratorInstitutionAdmins({
-      </body>
-    </html>
-  `.toString();
+  return PageLayout({
```
All `PageLayout` updates were merged from #11324.

nwalters512 commented:
This is awesome!! I haven't looked at code just yet, but I have some initial design thoughts/nits.

The tabs feel a little funny to me right now, they feel like they're sort of floating, un-anchored to anything:

<img width="2251" alt="Screenshot 2025-02-14 at 11 40 14" src="https://github.com/user-attachments/assets/f06bcf0c-bd55-4992-b1b5-6c504a8843e1" />

What do you think about something more like the existing layout instead, where they're clearly delineated with a separate background and anchored to the surrounding nav UI?

<img width="1952" alt="Screenshot 2025-02-14 at 11 40 04" src="https://github.com/user-attachments/assets/cdae3681-2ce9-4b17-80a2-28722ffbb88b" />

Making them unconditionally full width also has advantages for e.g. the course settings page, where if you click from Settings to Assessment Sets, the tabs jump around from narrower width to full width.

On many pages, the padding feels somewhat unbalanced:

<img width="2556" alt="Screenshot 2025-02-14 at 11 43 47" src="https://github.com/user-attachments/assets/5d56dc68-247d-4d64-aa86-4d45a328e5a5" />

I think I'd expect a more even `p-3` around the main content.

On the course nav (for discussion, cc @mwest1066 since I know he has opinions on this):

- I'd be inclined to push "Sharing" into the course "Settings" tabbed view.
- I'm wondering if we should pull "Staff" up into the sidebar and out of the "Settings" tabs? This is a thing that some instructors struggle to find. On the other hand, maybe it's a lot easier to tell instructors to just "click settings under your course, then click the staff tab".

Re: the assessment chooser that's visible at the top when navigating within a single assessment:

- The assessment chooser seems really large. Can we use a more typical font size for it?
- It also doesn't seem to work great if one has a ton of assessments. I don't think Bootstrap's dropdown component is responsive enough to scroll if the height of that exceeds the height of the viewport.
- How are the assessments in the chooser populated, is it by assessment set? If that's the case, it feels strange where people have opted to group their assessments by module. Maybe this is just the behavior of the existing assessment switcher in the navbar?
- Related to the above two points: I'm wondering if the assessment picker should be a richer modal that can show things like set/module grouping/ordering, the colored set/number badges, and so on. Maybe we save this for a future PR.
- The selector seems to do weird thing on narrow viewports:
  <img width="962" alt="Screenshot 2025-02-14 at 11 59 22" src="https://github.com/user-attachments/assets/9b5c286e-a78c-41c1-a2bc-acd3d6aaa14c" />

nwalters512 commented:
@coderabbitai pause

(things will be in a lot of flux for a while, I think)

nwalters512 commented:
I know this is likely still a work in progress, but I checked out the latest commit, it's looking even better! A few more UI notes (just writing them down as I find them so I don't forget):

- When one clicks into a question from the course questions page, it'd be great to keep the "Questions" item in the sidebar highlighted.
- When one clicks into a particular student's instance on the "Students" tab of an assessment, it'd be great to keep the "Students" tab highlighted.
- We seem to have lost some padding in at least one page that doesn't use the new sidebar (this is the student view assessments page):
  <img width="1370" alt="Screenshot 2025-02-14 at 17 15 46" src="https://github.com/user-attachments/assets/f584bd88-6451-47fa-9f78-15ecbb0a8d17" />
- There seems to be some double padding/margins at the bottom of some pages that are long enough to require scrolling. For instance, here's the bottom of the instructor assessment instance page:
  <img width="2547" alt="Screenshot 2025-02-14 at 17 18 05" src="https://github.com/user-attachments/assets/642ec42f-c747-4ac7-9431-75f2ed257749" />
- The icons in the sidebar are of varying widths, which causes some interesting alignment issues. I think we can use [Font Awesome's fixed width icons](https://docs.fontawesome.com/web/style/fixed-width) to resolve this.
  <img width="315" alt="Screenshot 2025-02-14 at 17 20 07" src="https://github.com/user-attachments/assets/29795c52-7972-45fc-8cde-3dc96baf9081" />

miguelaenlle commented:
@nwalters512 , I implemented your suggestions:

- Added background and fixed width to ContextNavigation:
![image](https://github.com/user-attachments/assets/9891bf60-13c1-44b3-b08d-7657a8446852)

- Balanced padding (consistent `p-3`, with `mb-4` on the bottom without bottom padding)
![image](https://github.com/user-attachments/assets/5d276c44-3cad-4d38-a5e7-762e5dcf30ea)

- Move Staff page to side nav; Move Sharing page to settings
![image](https://github.com/user-attachments/assets/dd9b9583-f07e-4fb5-87f7-6870f9bfd5e3)
![image](https://github.com/user-attachments/assets/f00b9ca8-797c-42f8-8145-e6b2dd0534ae)

- Reduced assessment selector size
![image](https://github.com/user-attachments/assets/adb0a024-b745-4170-a603-a301a12ebd8f)

- Fixed assessment selector content horizontal overflow issue - The assessment selector will now adjust its width based on the size of the assessment name
![image](https://github.com/user-attachments/assets/b509f149-ff0e-4eeb-bada-96cb218b9614)

- Fixed course, course instance, and assessment selector vertical overflow issues (scrollbar appears when the content is greater than `50vh` tall)
![image](https://github.com/user-attachments/assets/e96c2682-c9cb-4d24-8c7b-bf36354825e5)
![image](https://github.com/user-attachments/assets/69a2e704-5bb5-4d5b-92d1-b0828004db54)

- Maintained Questions side nav tab highlight when navigating to a course question
![image](https://github.com/user-attachments/assets/bb632904-beb7-4550-b955-9ed43ad9c63c)

- Maintained Students ContextNavigation nav tab highlight when navigating to an assessment instance
![image](https://github.com/user-attachments/assets/3bb036d5-ac8d-4423-929c-a2676899b5fb)

- Fixed incorrect padding issue on student view pages by reverting to default ContextNavigation behavior when the page does not use `PageLayout`:

https://github.com/PrairieLearn/PrairieLearn/blob/e61f776feef3c9053bcabc17856b7ebcb746f193/apps/prairielearn/src/components/Navbar.html.ts#L110-L116

![image](https://github.com/user-attachments/assets/c34fcf18-c2a1-47b3-b210-b1d87fbc357b)

![image](https://github.com/user-attachments/assets/a7271728-997f-491d-b16c-215f111f9d88)

- Fixed double padding issue on large pages with side nav:

<img width="1705" alt="image" src="https://github.com/user-attachments/assets/7bb8792c-a2d7-40c6-bc09-cac9d66cb40d" />

- Fixed side nav tab icon alignment with fixed-width Font Awesome icons:

![image](https://github.com/user-attachments/assets/88910c5f-1d7d-40a1-8855-debff174660c)

miguelaenlle commented:
> How are the assessments in the chooser populated, is it by assessment set?
The assessments in the chooser are populated by assessment set right now. This was the behavior of the existing assessment switcher in the navbar. I can adjust the query to populate it by set/module based on the course instance settings if needed 👍

miguelaenlle commented:
> I'm wondering if the assessment picker should be a richer modal that can show things like set/module grouping/ordering, the colored set/number badges, and so on

I agree that would be valuable. I'm inclined to save it for another PR to prevent this one from getting too big

nwalters512 commented:
> I can adjust the query to populate it by set/module based on the course instance settings if needed
> ...
> I'm inclined to save it for another PR to prevent this one from getting too big.

IMO changing the switcher to be more full-featured will be the way to go, and we should indeed save that for another PR! We could even land that as part of the existing assessment switcher, which would probably be pretty easy and would improve the experience for instructors in the meantime.

nwalters512 commented:
> When the ContextNavigation component has many tabs, it looks like the selected tab is hovering ... However, this is also how it looked prior to this PR. Any thoughts on adjusting this UI to address this or leaving it the same?

I think we should leave this as is for now since this PR doesn't make things any worse. It might be nice to change our approach in a future PR, either by making it horizontally scrollable or doing something like what GitHub does with repository tabs on narrow viewports.

> For bootstrap-tables with sticky-header components (e.g. the questions table), the header sticks to the top of the screen, going above the navbar:

Let me take a look at this, I can't immediately think of a fix but I'll play around and see if I can find one!

> If we navigate to a course instance, then to an assessment, then to a question, the Questions tab highlights in the Course section ... Let me know if you have any thoughts on this. I think it's fine, since questions are for the whole course, not just the course instance, but I'd be happy to change this behavior if needed.

This behavior is actually perfect for now. This is actually a step in the right direction for what we want to do in the long term: we generally consider it a mistake to have mounted all the course routes a second time under the course instance routes. We can't change this just yet, but we'll want to eventually, and the behavior you called out will make more sense when we do.

nwalters512 commented:
A few notes from reviewing this with Matt:

- Clicking the course "settings" in the sidebar then shows another tab called "Settings". Could we rename the tab label to "General", and maybe update the card title too? I acknowledge this will require a fork in behavior between the legacy nav and this new stuff (maybe that already exists).
- When the "Getting Started" option is visible in the sidebar, can we right-align the progress indicator to make it a little more visually distinct from the other items?
- For the course instance, can we shift "Access", "LTI", and "Billing" under a new tabbed navigation structure under the course instance "Settings" link in the sidebar? Ditto for then naming the main settings tab to "General".
- I wasn't able to immediately test this, but presumably "LTI 1.3" will also show up somewhere, and we'd like to make sure that's visible in the sidebar because for courses that have turned that on, they'll very frequently be going there to sync grades.
- For the course instance part of the sidebar, let's order them "Assessments", "Gradebook", "Files", "Settings".
- Matt says the assessment switcher dropdown doesn't look enough like a header, and that it kind of gets lost at the top of the page. His question: "can we make this look more like a header?" "I want it to look like a header when I don't need to interact with it, but then look like a switcher when I do." I have some ideas and might play around with this a bit on your branch if it's ok with you, @miguelaenlle.

miguelaenlle commented:
Thank you for the feedback, @nwalters512 and @mwest1066!

> * Clicking the course "settings" in the sidebar then shows another tab called "Settings". Could we rename the tab label to "General", and maybe update the card title too? I acknowledge this will require a fork in behavior between the legacy nav and this new stuff (maybe that already exists).
The side nav has a separate list of tabs from the ContextNavigation, so changing the side nav "Settings" tab label to "General" is very simple. Changing the `ContextNavigation` title from "Settings" to "General" is also quite straightforward since I made a function, `getNavPageTabs(hasEnhancedNavigation: boolean)`, that returns the nav page tabs for the new navigation and the old navigation.

> * When the "Getting Started" option is visible in the sidebar, can we right-align the progress indicator to make it a little more visually distinct from the other items?
I agree that would distinguish it more and improve the design.

> * For the course instance, can we shift "Access", "LTI", and "Billing" under a new tabbed navigation structure under the course instance "Settings" link in the sidebar? Ditto for then naming the main settings tab to "General".
That sounds good 👍 

> * I wasn't able to immediately test this, but presumably "LTI 1.3" will also show up somewhere, and we'd like to make sure that's visible in the sidebar because for courses that have turned that on, they'll very frequently be going there to sync grades.
This is accounted for, but I'll test it again:

https://github.com/PrairieLearn/PrairieLearn/blob/97f582af6af44a2f18513f8008d15bf299023bef/apps/prairielearn/src/components/SideNav.html.ts#L69-L76

> * For the course instance part of the sidebar, let's order them "Assessments", "Gradebook", "Files", "Settings".

Sounds good 👍 

> * Matt says the assessment switcher dropdown doesn't look enough like a header, and that it kind of gets lost at the top of the page. His question: "can we make this look more like a header?" "I want it to look like a header when I don't need to interact with it, but then look like a switcher when I do." I have some ideas and might play around with this a bit on your branch if it's ok with you, @miguelaenlle.

Got it. Feel free to add anything to the branch, @nwalters512 !

nwalters512 commented:
@miguelaenlle I pushed an update to the assessment selector styles in bc59371. Here's how it looks:

https://github.com/user-attachments/assets/b38defc6-2cc1-413f-a9e9-49726b0fd612

One thing to be aware of and check is how screen readers will experience these dropdowns. You may want to provide an `aria-label` like "Select assessment", "Select course instance", etc. so that folks using screen readers can tell what they are. On macOS, you can use [VoiceOver](https://support.apple.com/guide/voiceover/welcome/mac) to try things out for yourself.

miguelaenlle commented:
> @miguelaenlle I pushed an update to the assessment selector styles in [bc59371] 

Looks great, @nwalters512 ! I'll add `aria-label`s to selectors and also test them with `VoiceOver` - thank you for the suggestion.

miguelaenlle commented:
@nwalters512 and @mwest1066 , the changes are now live.

> Clicking the course "settings" in the sidebar then shows another tab called "Settings". Could we rename the tab label to "General", and maybe update the card title too? I acknowledge this will require a fork in behavior between the legacy nav and this new stuff (maybe that already exists).

I implemented this suggestion. 

When enhanced navigation is enabled, here's the Settings page:

![image](https://github.com/user-attachments/assets/d72f742f-b38a-476b-9b97-e4faba726dfe) 
- Renamed context nav tab label to General
- Updated card title

When enhanced navigation is not enabled:

![image](https://github.com/user-attachments/assets/5793849d-1516-4061-9403-62f7b0278d8e)

> When the "Getting Started" option is visible in the sidebar, can we right-align the progress indicator to make it a little more visually distinct from the other items?

I added the right alignment to the progress indicator:

![image](https://github.com/user-attachments/assets/88b4ada1-2bc9-4866-b04a-4a66e75aafd3)

> For the course instance, can we shift "Access", "LTI", and "Billing" under a new tabbed navigation structure under the course instance "Settings" link in the sidebar? Ditto for then naming the main settings tab to "General".

Here's the new settings page:

![image](https://github.com/user-attachments/assets/e44849c0-2215-437b-af7f-5a408badc67e)

> I wasn't able to immediately test this, but presumably "LTI 1.3" will also show up somewhere, and we'd like to make sure that's visible in the sidebar because for courses that have turned that on, they'll very frequently be going there to sync grades.

When `resLocals.lti13_enabled` is true, here's what the side nav looks like:

![image](https://github.com/user-attachments/assets/421b8afa-bb8f-43a9-bcd1-5424c19751bb)

> For the course instance part of the sidebar, let's order them "Assessments", "Gradebook", "Files", "Settings".

Here are the reordered tabs on the course instance part of the sidebar:

![Screenshot 2025-02-20 at 7 58 40 PM](https://github.com/user-attachments/assets/ed1870b7-e589-4d97-8154-eef0f6e10ff3)

mwest1066 commented:
Recording the discussion from the dev meeting today:

We thought that the current top-left hamburger menu is not discoverable because it doesn't visually look like it's part of the collapsable area. Instead we suggested moving the collapse icon to the top-right of the collapsible area and collapsing to an icon-only sidebar, which could then have the expand icon at the top. I think @nwalters512 already proposed an icon-only sidebar as the collapsed state, but we thought this should be implemented right away to allow for a discoverable collapse/expand.

There are some cool sidebar collapse/expand icons around the place:

* <img width="49" alt="image" src="https://github.com/user-attachments/assets/5febc021-0e9c-465b-ae3c-7e8c41bd12ab" />
* <img width="70" alt="image" src="https://github.com/user-attachments/assets/20d23890-e417-451f-a62c-882624d1bacd" />
* <img width="72" alt="image" src="https://github.com/user-attachments/assets/88356dfb-1858-4bd5-9fc0-e887aa092f7b" /> <img width="67" alt="image" src="https://github.com/user-attachments/assets/d3086a79-6051-4c44-9465-50b4aa9ecb53" />

But in Bootstrap the best options seem to be some kind of arrow?
* https://icons.getbootstrap.com/icons/chevron-left/
* https://icons.getbootstrap.com/icons/chevron-double-left/
* https://icons.getbootstrap.com/icons/caret-left-fill/
* https://icons.getbootstrap.com/icons/arrow-left-square/

nwalters512 commented on apps/prairielearn/src/ee/routers/institutionAdmin.ts:
```diff
@@ -1,14 +1,19 @@
+  const hasEnhancedNavigation = await features.enabled('enhanced-navigation', {
+    institution_id: req.params.institution_id,
+  });
+  res.locals.has_enhanced_navigation = hasEnhancedNavigation;
```
Tiniest of nits:

```suggestion
  res.locals.has_enhanced_navigation = hasEnhancedNavigation;

```

nwalters512 commented on apps/prairielearn/src/lib/authn.ts:
```diff
@@ -118,5 +118,17 @@ export async function loadUser(
 
+  res.locals.is_institution_administrator =
+    (res.locals.authn_is_administrator ||
+      (await sqldb.queryOptionalRow(
```
Since you're using `SELECT EXISTS ...`, this should always return a value, so you should be able to use `queryRow` here instead.

miguelaenlle replied:
Sounds good; I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/e8aee7bccff63fea945937d6e10f8151380e7bce/apps/prairielearn/src/lib/authn.ts#L121-L130

nwalters512 commented on apps/prairielearn/src/lib/course.ts:
```diff
@@ -192,3 +193,31 @@ export async function pullAndUpdateCourse({
 }
+
+/**
+ * Return the courses that the user has administrator access to.
```
"administrator" means a specific thing to us, and it isn't really what we're looking for here. I think "staff access" is more appropriate.

nwalters512 commented on apps/prairielearn/src/pages/instructorEffectiveUser/instructorEffectiveUser.ts:
```diff
@@ -1,4 +1,4 @@
-import { parseISO, isValid } from 'date-fns';
+import { isValid, parseISO } from 'date-fns';
```
While reorders like this are fine, they add unnecessary noise to PRs. I'd like to either a) figure out if we can make ESLint enforce this on all files, or b) disable this in your editor.

miguelaenlle replied:
The issue occurs when I remove an import, then add it back via an auto import.

For example, if I begin with the imports in this order:

![image](https://github.com/user-attachments/assets/a9fa9de9-b50d-44bd-bd2a-69ad9ac06449)

Then, I remove them, and re-import them via auto imports:

![image](https://github.com/user-attachments/assets/7ed2e855-bf87-4a41-b7b7-fa4b7751c380)

Regardless of the order I auto-import them in, they will be imported in this order:

![image](https://github.com/user-attachments/assets/7256fb34-942e-4e9c-a565-1bd700bedcf5)

Presumably, this is because of VSCode's built-in `Organize Imports` feature, which, when I trigger it manually:

![image](https://github.com/user-attachments/assets/cdb3dfbc-85b9-4928-9741-25aa57331a93)

Also flips the imports:

![image](https://github.com/user-attachments/assets/d519b68a-51b7-4527-88d9-23e7eb49bfaa)

I think the one of the following options could resolve the issue:
- Find a way to disable this behavior on my VSCode environment 
- Enable import alphabetization on ESLint:

![image](https://github.com/user-attachments/assets/6dc30729-64ba-4717-a1e7-63df03d62c9d)

(https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md)

Would you have any thoughts/recommendations on how to resolve the issue, @nwalters512 ?

nwalters512 replied:
I think getting ESLint to enforce this would be great! Would you be willing to open a separate PR so we can see what that would look like when applied to the whole codebase?

miguelaenlle replied:
I'd be happy to! I'll do that 👍

nwalters512 replied:
Thanks! I'll resolve this comment thread to tidy things up.

nwalters512 commented on apps/prairielearn/src/pages/instructorCourseAdminGettingStarted/instructorCourseAdminGettingStarted.html.ts:
Nit, consider lifting this to a small new PR to keep this one more focused.

miguelaenlle replied:
I opened a new PR that implements `PageLayout` for the getting started page: https://github.com/PrairieLearn/PrairieLearn/pull/11473

nwalters512 replied:
🎉 the other PR has been merged.

nwalters512 commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,332 @@
+      tabLabel: 'Settings',
+    },
+  ],
+  course_admin: [
```
Tiny nit: I'd be inclined to put this before `instance_admin` to reflect the fact that it comes higher in the navbar.

nwalters512 commented on apps/prairielearn/src/lib/navPageTabs.ts:
```diff
@@ -0,0 +1,478 @@
+ * @param hasEnhancedNavigation - Flag indicating if the enhanced navigation feature is enabled
+ * @returns Navigation page tabs and their configurations
+ */
+export function getNavPageTabs(hasEnhancedNavigation: boolean) {
```
I'd make the same change here as I suggested re: using `satisfies` to avoid losing type information.

IIRC this currently makes the assumption that each "nav page" can have a single set of tabbed pages, basically the "settings" hierarchy. For instance, a course couldn't have a "settings" sidebar item with tabs "foo" and "bar", and a separate "sync" sidebar item with tabs "repository" and "images". Am I understanding that right? As I type this out, maybe you'd just have to model that with separate nav pages, e.g. an `instance_admin_settings` nav page and another `instance_admin_sync` page.

miguelaenlle replied:
> I'd make the same change here as I suggested re: using satisfies to avoid losing type information.

Sounds good; I made this change, and removed all corresponding `undefined` checks.

 > IIRC this currently makes the assumption that each "nav page" can have a single set of tabbed pages, basically the "settings" hierarchy
 
Correct. Each nav page on the side nav has a set of (horizontally) tabbed pages.

> For instance, a course couldn't have a "settings" sidebar item with tabs "foo" and "bar", and a separate "sync" sidebar item with tabs "repository" and "images".

That is correct. The `course_admin` page has one set of pages that can be displayed, so the "settings" sidebar item can have tabs "foo" and "bar", but the separate "sync" sidebar item must have "foo" and "bar" as well.

Whether or not the `ContextNavigation` tabs are displayed is controlled in `PageLayout`. When `showContextNavigation` is true, the tabs are displayed.

https://github.com/PrairieLearn/PrairieLearn/blob/732400ef7f849390b09ed944b64574bee71e57c2/apps/prairielearn/src/components/PageLayout.html.ts#L62-L88

> As I type this out, maybe you'd just have to model that with separate nav pages, e.g. an instance_admin_settings nav page and another instance_admin_sync page.

This would be the way to model it with the current implementation. If `instance_admin_settings` and `instance_admin_sync` should be under the course instances section of the side nav, new `SideNavTabInfo` objects would be added to `sideNavPagesTabs.instance_admin`:

https://github.com/PrairieLearn/PrairieLearn/blob/732400ef7f849390b09ed944b64574bee71e57c2/apps/prairielearn/src/components/SideNav.html.ts#L105-L144

And for them,`activePages` would be set to `["instance_admin_settings"]` and `["instance_admin_sync"]` respectively.

nwalters512 replied:
Thanks for explaining that! Glad to know this is possible if we need it.

nwalters512 commented:
@mwest1066 re: an icon-only collapsed state, Miguel and I discussed that and agreed that we'd update #11425 to include that before landing. We still intend on landing this PR (#11279) without any collapsing nav to avoid this PR becoming too large.

miguelaenlle commented on apps/prairielearn/src/components/AssessmentNavigation.html.ts:
```diff
@@ -0,0 +1,48 @@
+            return html`
+              <a
+                class="dropdown-item ${`${assessment.id}` === `${a.id}` ? 'active' : ''}"
+                aria-current="${`${assessment.id}` === `${a.id}` ? 'page' : ''}"
```
I tested the dropdown in VoiceOver, and the screen reader functions slightly differently with and without the `aria-current` attribute. 

With the attribute:
![Screenshot 2025-02-27 at 5 41 33 PM](https://github.com/user-attachments/assets/8c0c8933-5ed3-4199-adea-b9ef53d950bd)

Without the attribute:
![Screenshot 2025-02-27 at 5 43 42 PM](https://github.com/user-attachments/assets/8d9356e4-bb69-4914-90b3-c59cd18dfaf4)

Thus, I'm inclined to keep the attribute, but let me know your thoughts @nwalters512

nwalters512 replied:
Right, we definitely want it if it does reflect the current page. My question was, for links that aren't current page, if it make a difference to use `aria-current=""` vs. just omitting the attribute entirely. It seems like it doesn't make a difference in your testing, so I'm happy to keep it.

FWIW, I jsut find the following on MDN(https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-current), which confirms that this is fine:

> If the attribute is not present, is an empty string, is present with no value, or is set to `aria-current="false"` it is not exposed to the user.

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -14,13 +14,13 @@ export function Navbar({
   navSubPage?: NavSubPage;
   navbarType?: NavbarType;
-  marginBottom?: boolean;
+  isInPageLayout?: boolean;
```
I added a comment. `isInPageLayout` is set to true when the `Navbar` is rendered in the `PageLayout` component. This is to ensure that enhanced navigation features are only present on pages that use PageLayout.

mwest1066 commented:
> @mwest1066 re: an icon-only collapsed state, Miguel and I discussed that and agreed that we'd update #11425 to include that before landing. We still intend on landing this PR (#11279) without any collapsing nav to avoid this PR becoming too large.

Oh sorry, I commented on the wrong PR! I meant to post this comment on #11425. The plan sounds perfect to me.

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.html.ts:
```diff
@@ -51,7 +51,6 @@ export function InstructorAiGenerateDraftEditor({
                 navPage: 'course_admin',
                 navSubPage: 'questions',
                 resLocals,
-                marginBottom: false,
```
Got it, thanks for the heads up @nwalters512 . I added `marginBottom` back to the Navbar. If enhanced navigation is disabled or the page doesn't use `PageLayout`, when `marginBottom` is true, the div containing the `ContextNavigation` has `mb-3` when `marginBottom` is true:

https://github.com/PrairieLearn/PrairieLearn/blob/e8aee7bccff63fea945937d6e10f8151380e7bce/apps/prairielearn/src/components/Navbar.html.ts#L110-L124

`marginBottom` also still defaults to `true`.

https://github.com/PrairieLearn/PrairieLearn/blob/e8aee7bccff63fea945937d6e10f8151380e7bce/apps/prairielearn/src/components/Navbar.html.ts#L12-L24

miguelaenlle commented on apps/prairielearn/src/middlewares/authzCourseOrInstance.ts:
```diff
@@ -65,9 +67,25 @@ export async function authzCourseOrInstance(req: Request, res: Response) {
+    userId: res.locals.authn_user.user_id,
+    isAdministrator: res.locals.is_administrator,
+    includeExampleCourse: res.locals.is_administrator || config.devMode,
+  });
```
Got it; thank you for letting me know about this @nwalters512 ! 

I removed the queries to retrieve the courses and course instances in `authzCourseOrInstance`, and implemented `hx-get` requests in the Side Nav for course selection. I used the existing HTMX endpoints and very similar code to what's in the current nav bar:

https://github.com/PrairieLearn/PrairieLearn/blob/3eca142be1aed54873369105f34fe735f01fb603/apps/prairielearn/src/components/SideNav.html.ts#L245-L272

https://github.com/PrairieLearn/PrairieLearn/blob/3eca142be1aed54873369105f34fe735f01fb603/apps/prairielearn/src/components/SideNav.html.ts#L244-L281

I also added `aria-label`s to these dropdowns, identical to those on the current nav bar.

miguelaenlle replied:
I also added `mergeParams: true` to the router in `navbarCourseSwitcher`:

https://github.com/PrairieLearn/PrairieLearn/blob/c6803533dee4fc5236f185600dd47da45fd39d24/apps/prairielearn/src/pages/navbarCourseSwitcher/navbarCourseSwitcher.ts#L8-L11

This is because, without it, `navbarCourseSwitcher` doesn't retrieve the `course_id` param, so the current page isn't highlighted.

Before the fix (QA 101 is not highlighted):

![image](https://github.com/user-attachments/assets/7e69ac31-bcf7-4182-812f-ba3b226b85be)

After fix (QA 101 is highlighted):

![image](https://github.com/user-attachments/assets/d0c8d594-c932-430d-b416-dfcfe9dc1e6d)

miguelaenlle replied:
Lastly, I added `aria-current` to the `navbarCourseSwitcher`:

https://github.com/PrairieLearn/PrairieLearn/blob/c6803533dee4fc5236f185600dd47da45fd39d24/apps/prairielearn/src/pages/navbarCourseSwitcher/navbarCourseSwitcher.html.ts#L18-L22

nwalters512 replied:
Perfect 🙏

nwalters512 commented on apps/prairielearn/src/components/NavbarContext.html.ts:
```diff
@@ -421,12 +14,23 @@ export function ContextNavigation({
+          nav 
+          nav-tabs 
+          pl-nav-tabs-bar 
+          ${showSideNav ? 'pt-3' : 'pt-2'} 
```
Nit, IMO we should retain the smaller vertical spacing to avoid excessive use of whitespace.

```suggestion
          pt-2
```

miguelaenlle replied:
I agree with removing this since the removal reduces whitespace and simplifies the code a bit. Here's what pages with `ContextNavigation` look like with `pt-2` and the side nav:

![image](https://github.com/user-attachments/assets/3cdbe557-6bde-49e2-9afc-a75a7f4a7415)

nwalters512 commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,333 @@
+            </div>
+          </div>
+        </div>
+        ${courseSideNavPageTabs.map((tabInfo) =>
+          SideNavLink({
+            resLocals,
+            navPage: page,
+            navSubPage: subPage,
+            tabInfo,
+          }),
+        )}
```
Nit: this is nested inside the `<div class="dropdown">`, which is probably a mistake.

miguelaenlle replied:
Yup, that was a mistake. I moved this code out of the `dropdown` div:

https://github.com/PrairieLearn/PrairieLearn/blob/2333942dcc18ae922eff3d93f976fc0f2b9cd8fb/apps/prairielearn/src/components/SideNav.html.ts#L248-L285

nwalters512 replied:
Looks good, thanks!

nwalters512 commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,333 @@
+  return html`
+    <div class="side-nav-section-header">Course</div>
+    <div class="side-nav-group mb-3">
+      <div class="dropdown">
```
Nit, I'd like to add a small bottom margin to the dropdowns in the sidebar (probably `mb-2`). Without that, IMO the selected page looks weird butted up against the dropdown trigger:

<img width="313" alt="Screenshot 2025-03-04 at 10 05 03" src="https://github.com/user-attachments/assets/9a2bfec8-4026-47ef-8f52-158ec4e789e3" />

miguelaenlle replied:
I agree that it looks better with bottom margin below the dropdowns. Here's what they look like with the padding:

![Screenshot 2025-03-04 at 4 03 58 PM](https://github.com/user-attachments/assets/7fa8058d-fda1-466b-9b6b-6c424478be0b)

![Screenshot 2025-03-04 at 4 04 03 PM](https://github.com/user-attachments/assets/3d8f4abe-60e5-4cae-a403-1beb263cafbd)

nwalters512 replied:
Looks great! 🎉

nwalters512 commented on apps/prairielearn/assets/scripts/bootstrap-table-sticky-header.js:
Ah, I missed this on https://github.com/PrairieLearn/PrairieLearn/pull/11438 and maybe never explicitly called it out: my intent was for this to land on `master` independently of this PR so that it wouldn't make this PR even larger to review. Something to keep in mind for future PRs.

miguelaenlle replied:
Apologies for merging #11438 directly into this branch instead of `master`. In the future, I'll make sure to confirm which branch the PR should merge into, particularly in situations like this where related branches either merge into the `new-navigation-menus` branch or the `master` branch.

nwalters512 commented on apps/prairielearn/src/lib/course.ts:
```diff
@@ -192,3 +193,31 @@ export async function pullAndUpdateCourse({
   return { jobSequenceId: serverJob.jobSequenceId, jobPromise };
 }
+
+/**
+ * Return the courses that the user has staff access to.
+ */
+export async function selectInstructorCourses({
+  userId,
+  isAdministrator,
+  includeExampleCourse,
+}: {
+  userId: string;
+  isAdministrator: boolean;
+  /**
+   * Example courses are only shown to users who are either instructors of
+   * at least one other course, or who are admins. They're also shown
+   * unconditionally in dev mode.
+   */
+  includeExampleCourse: boolean;
+}) {
+  return await sqldb.queryRows(
+    sql.select_instructor_courses,
+    {
+      user_id: userId,
+      is_administrator: isAdministrator,
+      include_example_course: includeExampleCourse,
+    },
+    InstructorCourseSchema,
+  );
+}
```
Nit: AFAICT this function isn't actually used by any part of this PR other than the homepage, where the query originally lived. Would you mind reverting this change to cut down on the scope of this PR?

nwalters512 replied:
Thanks!

miguelaenlle commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,333 @@
+          ${!resLocals.authz_data.overrides
+            ? html`
+                hx-get="/pl/navbar/course/${resLocals.course.id}/switcher"
+                hx-trigger="show.bs.dropdown once delay:200ms"
```
Sounds good, thank you for the suggestion @nwalters512! I added this change to the course and course instance switchers on the side nav. 

On the (original) navbar, `mouseover once` and `focus once` are already triggers:

https://github.com/PrairieLearn/PrairieLearn/blob/c335e933ec0760d6bc8e3fd9595d18f4e99ee94a/apps/prairielearn/src/components/Navbar.html.ts#L778-L793

But they are not on the navbar course instance switcher:

https://github.com/PrairieLearn/PrairieLearn/blob/c335e933ec0760d6bc8e3fd9595d18f4e99ee94a/apps/prairielearn/src/components/Navbar.html.ts#L871-L883

I didn't add `mouseover once, focus once,` to the course instance switcher to limit the scope of this PR, but let me know if I should add it here as well. Alternatively, I can open a separate PR for it.

nwalters512 replied:
A separate PR adding it to the old navbar would be great! I think it's worth it to keep the old navbar working as well as possible until we can completely replace it with the new nav.

miguelaenlle replied:
Sounds great! I'll work on that 👍

nwalters512 replied:
Thanks! I'll resolve this thread then.

nwalters512 commented on apps/prairielearn/src/components/SideNav.html.ts:
```diff
@@ -0,0 +1,340 @@
+            }),
+          )}
+        </div>
+      </div>
```
This was since resolved.

nwalters512 commented on apps/prairielearn/src/lib/navPageTabs.ts:
```diff
@@ -0,0 +1,451 @@
+            iconClasses: 'fas fa-cog',
+            tabLabel: 'Settings',
+          },
+        ],
```
This is intentional, the other "tabs" are now shown in the sidebar.

nwalters512 commented on apps/prairielearn/src/lib/course.ts:
```diff
@@ -192,3 +193,31 @@ export async function pullAndUpdateCourse({
   return { jobSequenceId: serverJob.jobSequenceId, jobPromise };
 }
+
+/**
+ * Return the courses that the user has staff access to.
+ */
+export async function selectInstructorCourses({
```
Let's make sure to remove this function and the corresponding SQL since this function is now unused.

miguelaenlle replied:
I removed the `selectInstructorCourses` function and the corresponding SQL.

nwalters512 replied:
All good! Thanks for cleaning that up.

miguelaenlle commented on apps/prairielearn/src/components/NavbarContext.html.ts:
```diff
@@ -16,7 +16,10 @@ export function ContextNavigation({
+  const navPageTabs = navPagesTabs[navPage]; 
+  
+  // Some navPages do not have tabs
+  if (!navPageTabs) return '';
```
I added this because on certain pages, the `navPage` has no associated tabs, but `NavbarContext` is displayed. This results in the `ContextNavigation` component being visible but empty. One example would be the Course Requests page:

![Screenshot 2025-03-04 at 5 21 46 PM](https://github.com/user-attachments/assets/9f3edb45-3d45-4afb-9bbc-17bdbbd57d03)

This issue is now fixed:

![Screenshot 2025-03-04 at 5 22 14 PM](https://github.com/user-attachments/assets/ddc6eef7-b8a7-475c-830b-78c52336897e)

nwalters512 replied:
Very subtle, nice catch!

nwalters512 reviewed (approved):
My primary concern at this point with this PR is ensuring that we don't break things for people who don't have the `enhanced-navigation` feature flag turned on. I clicked through as many of our pages as I could with the feature flag disabled and confirmed that things are working as expected.

This is a really really good start! I'm very excited to start turning this on for people and getting some feedback.
